### PR TITLE
Task/Implement User layer

### PR DESCRIPTION
**Pull request by Nikname2303**

### CODE REVIEW

File: src/main/java/com/ua/accommodation/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+@RequestMapping("/api/users")
+@RequiredArgsConstructor
+public class UserController {
+    private final UserService userService;
+
+    @GetMapping("/me")
+    public UserResponseDto getUser(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return userService.getUser(user.getId());
+    }
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/controller/UserController.java:28]:
```suggestion
    public UserResponseDto getUser(@AuthenticationPrincipal User user) {
        return userService.getUser(user.getId());
    }
```

File: src/main/java/com/ua/accommodation/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    @PutMapping("{id}/role")
+    public UserResponseDto updateRoles(@PathVariable Long id,
+                                       @RequestBody UserUpdateRoleDto updateRoleDto) {
+        return userService.updateRoles(id, updateRoleDto);
+    }
+
+    @PatchMapping("/me")
+    public UserResponseDto updateProfile(Authentication authentication,
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/controller/UserController.java:37]: Same here

File: src/main/java/com/ua/accommodation/dto/user/UserUpdateProfileDto.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.ua.accommodation.dto.user;
+
+import lombok.Data;
+
+@Data
+public class UserUpdateProfileDto {
+    private String email;
+    private String firstName;
+    private String lastName;
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/dto/user/UserUpdateProfileDto.java:9]: Please add validation, and lombok.validator dependency

File: src/main/java/com/ua/accommodation/dto/user/UserUpdateRoleDto.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.ua.accommodation.model.Role;
+import java.util.Set;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+@Data
+public class UserUpdateRoleDto {
+    private Long id;
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/dto/user/UserUpdateRoleDto.java:11]: Why do you need id here? You accept user id as PathVariable and set of roles as RequestBody

File: src/main/java/com/ua/accommodation/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.ua.accommodation.repository;
+
+import com.ua.accommodation.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/repository/UserRepository.java:7]: this annotation not necessary if you implement JPARepository

File: src/main/java/com/ua/accommodation/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto getUser(Long userId) {
+        User user = userRepository.findById(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find user with id: " + userId)
+        );
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/service/impl/UserServiceImpl.java:24]: Extract method, because this code duplicates on the next method

**Review by slizko1 — CHANGES_REQUESTED** [review]
Some improvements are needed

File: src/main/java/com/ua/accommodation/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                                       @RequestBody @Valid UserUpdateRoleDto updateRoleDto) {
+        return userService.updateRoles(id, updateRoleDto);
+    }
+
+    @PatchMapping("/me")
+    public UserResponseDto updateProfile(@AuthenticationPrincipal User user,
+                                         @RequestBody
+                                         @Valid UserUpdateProfileDto updateProfileDto) {
+        return userService.updateProfile(user.getId(), updateProfileDto);
+    }
```
Comment (slizko1) [inline src/main/java/com/ua/accommodation/controller/UserController.java:41]:
```suggestion
    public UserResponseDto updateProfile(
            @AuthenticationPrincipal User user,
            @RequestBody @Valid UserUpdateProfileDto updateProfileDto) {
        return userService.updateProfile(user.getId(), updateProfileDto);
    }
```
I think in that way it looks better

**Review by slizko1 — APPROVED** [review]
GJ! One comment was added

**Review by vikkushnir — APPROVED** [review]
LGTM!

**Pull request merged**